File: app/src/main/java/com/bykea/pk/partner/ui/withdraw/WithdrawalPaymentMethodsAdapter.java
package com.bykea.pk.partner.ui.withdraw;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bykea.pk.partner.R;
import com.bykea.pk.partner.dal.source.remote.data.WithdrawPaymentMethod;

import java.util.ArrayList;
import java.util.List;

public class WithdrawalPaymentMethodsAdapter extends RecyclerView.Adapter<WithdrawalPaymentMethodsAdapter.ViewHolder> {
    private final ArrayList<WithdrawPaymentMethod> withdrawPaymentMethods;
    private final WithdrawalViewModel viewModel;
    private int lastSelectedPosition;

    public WithdrawalPaymentMethodsAdapter(ArrayList<WithdrawPaymentMethod> withdrawPaymentMethods,
                                           WithdrawalViewModel viewModel) {
        this.withdrawPaymentMethods = withdrawPaymentMethods;
        this.viewModel = viewModel;
        this.lastSelectedPosition = 0;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater =
                LayoutInflater.from(parent.getContext());
        return new ViewHolder(layoutInflater.inflate(R.layout.adapter_withdraw_methdods, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.mCnicTextView.setText(viewModel.getDriverCnicNumber());
        holder.mFeesTextView.setText(withdrawPaymentMethods.get(position).getDescription());
        holder.mCheckImageView.setVisibility(withdrawPaymentMethods.get(position).isSelected() ?
                View.VISIBLE : View.INVISIBLE);
    }

    @Override
    public int getItemCount() {
        return withdrawPaymentMethods.size();
    }

    public void notifyMethodsChanged(List<WithdrawPaymentMethod> it) {
        withdrawPaymentMethods.clear();
        withdrawPaymentMethods.addAll(it);
        notifyDataSetChanged();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final TextView mCnicTextView;
        private final TextView mFeesTextView;
        private final ImageView mCheckImageView;

        public ViewHolder(View v) {
            super(v);
            mCnicTextView = v.findViewById(R.id.nic_val_textview);
            mFeesTextView = v.findViewById(R.id.fees_val_textview);
            mCheckImageView = v.findViewById(R.id.checkView);
            v.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();
            WithdrawPaymentMethod object = withdrawPaymentMethods.get(position);
            if (!object.isSelected()) {
                object.setSelected(true);
                viewModel.setSelectedPaymentMethod(object);
                withdrawPaymentMethods.get(lastSelectedPosition).setSelected(false);
                lastSelectedPosition = position;
                notifyDataSetChanged();
            }
        }
    }
}

File: app/src/main/java/com/bykea/pk/partner/ui/withdraw/WithdrawalPaymentMethodsAdapter.kt
package com.bykea.pk.partner.ui.withdraw

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.bykea.pk.partner.R
import com.bykea.pk.partner.dal.source.remote.data.WithdrawPaymentMethod
import com.bykea.pk.partner.databinding.AdapterWithdrawMethdodsBinding
import java.util.*

/**
 * this method will provide the datasource to the
 * payment method recyclerview on {WithdrawalActivity}
 *
 * @param withdrawPaymentMethods provides the list of payment methods to the datasource
 * @param viewModel will provide the business logic to the datasource
 *
 * @author Arsal Imam
 */
class WithdrawalPaymentMethodsAdapter(private val withdrawPaymentMethods: ArrayList<WithdrawPaymentMethod>,
                                      private val viewModel: WithdrawalViewModel?) : RecyclerView.Adapter<WithdrawalPaymentMethodsAdapter.ViewHolder>() {

    /**
     * last selected position to maintain checkbox selection
     */
    private var lastSelectedPosition: Int = 0

    /**
     * constructor of the class
     */
    init {
        this.lastSelectedPosition = 0
    }

    /**
     * {@inheritDoc}
     *
     * This method will invoke by the system to create view for row
     *
     * @param parent in which view will create
     * @param viewType this will be managed from {@link #getItemViewType(position: Int):Int}
     *
     * @return viewHolder instance to create row
     */
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val binding: AdapterWithdrawMethdodsBinding = DataBindingUtil.inflate(layoutInflater,
                R.layout.adapter_withdraw_methdods, parent, false)
        return ViewHolder(binding, binding.root)
    }

    /**
     * this method will bind view with data
     *
     * @param holder this was created in the {@link #onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder}
     * @param position of the row
     */
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.updateViewHolder(viewModel, withdrawPaymentMethods[position])
    }

    /**
     * return total counts of the row to be created
     *
     * @return count of rows
     */
    override fun getItemCount(): Int {
        return withdrawPaymentMethods.size
    }

    /**
     * this method can be used to update the datasource's data
     *
     * @param it updated data through which new rows needs to be created
     */
    fun notifyMethodsChanged(it: List<WithdrawPaymentMethod>) {
        withdrawPaymentMethods.clear()
        withdrawPaymentMethods.addAll(it)
        notifyDataSetChanged()
    }

    /**
     * viewHolder of the datasource which will hold views
     * @param v view that needs to be hold
     */
    inner class ViewHolder(
            val binder: AdapterWithdrawMethdodsBinding, v: View)
        : RecyclerView.ViewHolder(v), View.OnClickListener {

        /**
         * constructor of the class
         */
        init {
            v.setOnClickListener(this)
        }

        /**
         * click event when user click on the row
         *
         * @param view on which user tapped
         */
        override fun onClick(view: View) {
            val position = adapterPosition
            val `object` = withdrawPaymentMethods[position]
            if (!`object`.isSelected) {
                `object`.isSelected = true
                viewModel?.selectedPaymentMethod = `object`
                withdrawPaymentMethods[lastSelectedPosition].isSelected = false
                lastSelectedPosition = position
                notifyDataSetChanged()
            }
        }

        /**
         * this method will bind data with viewholder from {@link #onBindViewHolder(holder: ViewHolder, position: Int)}
         *
         * @param withdrawPaymentMethod object of the row
         * @param viewModel responsible viewModel
         */
        fun updateViewHolder(viewModel: WithdrawalViewModel?,
                             withdrawPaymentMethod: WithdrawPaymentMethod) {
            binder.nicValTextview.text = viewModel?.getDriverCnicNumber()
            binder.feesValTextview.text = withdrawPaymentMethod.description
            binder.checkView.visibility = if (withdrawPaymentMethod.isSelected)
                View.VISIBLE
            else
                View.INVISIBLE
        }
    }
}